### PR TITLE
App: Dependency Injection extensions

### PR DESCRIPTION
**Pull request by dstenroejl**
## Description

- [x] Removed use of "Microsoft.VisualStudio.Threading.Analyzers".
- [x] Upgrade to .NET 8

- [x] Implement and test reusable dependency injection extensions for App's (Common):
  - [x] Health Checks (TryAdd)
  - [x] Noda Time

- [x] Implement and test reusable dependency injection extensions for Web App's:
  - [x] Telemetry/logging using Application Insights
  - [x] Health Checks
  - [x] Authentication / Authorization
  - [x] Api Versioning
  - [x] Open Api

- [x] Implement and test reusable dependency injection extensions for Azure Function App's:
  - [x] Telemetry/logging using Application Insights
  - [x] Health Checks

Part of:
https://app.zenhub.com/workspaces/mandalorian-5fd22a1a59e4740018f5ab5a/issues/gh/energinet-datahub/team-mandalorian/237

## Quality

- [x] Documentation is updated
- [x] Release notes are updated
- [x] Package version is updated
- [x] Public types and methods are documented
- [x] Tests are implemented and executed locally

### CODE REVIEW

File: source/App/documents/documentation.md [hunk 1]
```diff
@@ -1,12 +1,182 @@
...
+
+1) Add `Program.cs` with the following content
+
+   ```cs
+   var host = new HostBuilder()
+       .ConfigureFunctionsWorkerDefaults()
+       .ConfigureServices((context, services) =>
+       {
+           // Common
+           services.AddApplicationInsightsForIsolatedWorker(subsystemName: "MySubsystem");
```
[the remark this answers is not in the document]
Comment (dstenroejl) [reply source/App/documents/documentation.md:57]: Yes. It's just added as a value to the custom property Subsystem when emitting telemetry (logs, request, etc.)

File: .github/workflows/app-common-bundle-publish.yml [hunk 2]
```diff
@@ -70,6 +70,7 @@ jobs:
         with:
           use_azure_functions_tools: "true"
           azure_functions_core_tools_version: 4.0.5455
+          dotnet_version: 8.0.200
```
Comment (defectiveAi) [inline .github/workflows/app-common-bundle-publish.yml:73]: Not sure why this is needed. This does not match the newest version, and a .net 8 package is built without specifing a version.

Comment (dstenroejl) [reply .github/workflows/app-common-bundle-publish.yml:73]: This version decides the "tools versions" (SDK) used when executing command lines. e.g. `dotnet build`. 

Ideally it should be configured using a `global.json` file, then we could ensure it's the same locally and on the build agent. But I'm just not up for handling that change at the moment.

I can however see the default value has now been updated by Outlaws (previously it was 7.x), so we can avoid overriding it now.

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs [hunk 3]
```diff
@@ -62,7 +65,8 @@ public static void UseUserMiddleware<TUser>(this IApplicationBuilder builder)
     /// <param name="externalMetadataAddress">The address of OpenId configuration endpoint for the external token, e.g. https://{b2clogin.com/tenant-id/policy}/v2.0/.well-known/openid-configuration.</param>
     /// <param name="internalMetadataAddress">The address of OpenId configuration endpoint for the internal token, e.g. https://{market-participant-web-api}/.well-known/openid-configuration.</param>
     /// <param name="backendAppId"></param>
-    public static void AddJwtBearerAuthentication(
+    [Obsolete("Should only be used for testing.")]
```
Comment (defectiveAi) [inline source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs:68]: Add name of correct function to obsolete message: AddJwtBearerAuthenticationForWebApp

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs [hunk 4]
```diff
@@ -15,34 +15,35 @@
...
+    internal static bool DisableHttpsConfiguration { get; set; }
 
-    public static void UseUserMiddleware<TUser>(this IApplicationBuilder builder)
-        where TUser : class
-    {
-        builder.UseMiddleware<UserMiddleware<TUser>>();
-    }
-
-    public static void AddUserAuthentication<TUser, TUserProvider>(this IServiceCollection services)
+    // TODO: Add description
```
Comment (defectiveAi) [inline source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs:45]: A todo. Maybe not even have a description, as method name is clear already.

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthorizationExtensions.cs [hunk 5]
```diff
@@ -14,12 +14,19 @@
...
+namespace Energinet.DataHub.Core.App.WebApp.Extensions.DependencyInjection;
 
+/// <summary>
+/// Extension methods for <see cref="IServiceCollection"/>
+/// that allow adding authorization services to an ASP.NET Core app.
+/// </summary>
 public static class AuthorizationExtensions
 {
-    public static void AddPermissionAuthorization(this IServiceCollection services)
+    // TODO: Add description
```
Comment (defectiveAi) [inline source/App/source/WebApp/Extensions/DependencyInjection/AuthorizationExtensions.cs:25]: Same comment as before about description; writing 'Adds permission authorization to a WebAPI' does not add value.

File: source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    /// The address of MitID configuration endpoint for the external token.
+    /// </summary>
+    public string MitIdExternalMetadataAddress { get; set; } = string.Empty;
+
+    /// <summary>
+    /// The address of OpenId configuration endpoint for the external token, e.g. https://{b2clogin.com/tenant-id/policy}/v2.0/.well-known/openid-configuration.
+    /// </summary>
+    public string ExternalMetadataAddress { get; set; } = string.Empty;
+
+    public string BackendBffAppId { get; set; } = string.Empty;
```
Comment (defectiveAi) [inline source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs:32]: "The id of the backend-for-frontend app, which the access tokens are issue for. The audience claim in the tokens are expected to have this value."

File: source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.Core.App.WebApp.Extensions.Options;
+
+// TODO: Are we sure about this name? Microsoft has a class with the same name, but they also have a JwtBearerAuthenticationOptions class.
+public class AuthenticationOptions
+{
+    public const string SectionName = "Authentication";
```
Comment (defectiveAi) [inline source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs:20]: There could be confusion, as we use auth for much stuff. UserAuthenticationOptions maybe?

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs (hunk 3 above)
Comment (dstenroejl) [reply source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs:68]: Updated according to suggestion

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs (hunk 4 above)
Comment (dstenroejl) [reply source/App/source/WebApp/Extensions/DependencyInjection/AuthenticationExtensions.cs:45]: Removed TODO

File: source/App/source/WebApp/Extensions/DependencyInjection/AuthorizationExtensions.cs (hunk 5 above)
Comment (dstenroejl) [reply source/App/source/WebApp/Extensions/DependencyInjection/AuthorizationExtensions.cs:25]: Removed TODO

File: source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs (hunk 7 above)
Comment (dstenroejl) [reply source/App/source/WebApp/Extensions/Options/AuthenticationOptions.cs]: Renamed to 'UserAuthenticationOptions'

**Review by RasmusGraabaek — APPROVED** [review]

**Pull request merged**